geoSegment.js

```diff
@@ -0,0 +1,353 @@
+      return;
+    }
+
+    // Does ScheduledSession have downloaded superEvent?
+    const scheduledSessionSuperEventId = scheduledSessionItem.data.superEvent;
+    const sessionSeriesFilePath = getSessionSeriesFilePath(scheduledSessionSuperEventId);
+    if (!SESSION_SERIES_FILE_PATHS.includes(sessionSeriesFilePath)) {
+      return;
+    }
+
+    // Get SessionSeries
+    const sessionSeries = await fs.readJson(sessionSeriesFilePath);
+
+    // Link ScheduledSession and SessionSeries, and write
+    await linkScheduledSessionAndSessionSeriesAndWrite(scheduledSessionItem.data, sessionSeries);
+  });
+}
+
+async function downloadScheduledSessions(firehoseBaseUrl, firehoseApiKey) {
+  console.log('downloadAndWriteScheduledSessions() - starting');
+  // Get ScheduledSessions from Firehose
+  let nextUrl = `${firehoseBaseUrl}scheduled-sessions`;
+  let backoffTimeInSeconds = RETRY_BACKOFF_MIN;
+  while (true) {
+    let nextNextUrl;
+    let items;
+    try {
+      ({ nextNextUrl, items } = await getFirehosePage(nextUrl, firehoseApiKey));
+    } catch (error) {
+      if (error instanceof Http404Error) {
+        console.error(`ERROR: URL ("${error.url}") not found. Please check the value for <rpde-endpoint>.`);
+        process.exit(1);
+      }
+      const loggableError = (error && error.isAxiosError)
+        ? loggableAxiosError(error)
+        : error;
+      console.warn(`WARN: Retrying [page: "${nextUrl}"] due to error:`, loggableError);
+      if (backoffTimeInSeconds > RETRY_BACKOFF_MAX) {
+        console.error('ERROR: Cannot download Firehose pages, backoff limit reached, terminating script');
+        process.exit(1);
+      }
+      await wait(backoffTimeInSeconds * 1000);
+      backoffTimeInSeconds *= RETRY_BACKOFF_EXPONENTIATION_RATE;
+      continue;
+    }
+
+    // Process SessionSeries items
+    await processScheduledSessionItems(items);
+
+    // Have we reached the end of the feed?
+    if (nextNextUrl === nextUrl) {
+      break;
+    }
+
+    // If we haven't reached the end of the feed, carry on
+    nextUrl = nextNextUrl;
+    backoffTimeInSeconds = RETRY_BACKOFF_MIN;
+    await wait(MIN_REQUEST_DELAY_SECONDS * 1000);
+  }
+}
+
+(async () => {
+  console.log('geoSegment() - starting..');
+
+  // Load config file into memory
+  const { firehoseBaseUrl, firehoseApiKey, segments } = await fs.readJson(CONFIG_FILE_PATH);
+
+  // Create SessionSeries directory
+  await emptyOrMakeDirectory(SESSION_SERIES_DIRECTORY_PATH);
+
+  // Create output directory
+  await emptyOrMakeDirectory(OUTPUT_DIRECTORY_PATH);
+
+  // Create Segment directories
+  await createSegmentDirectories(segments);
+
+  // Download and write SessionSeries
+  await downloadSessionSeries(firehoseBaseUrl, firehoseApiKey, segments);
+
+  // Download and write ScheduledSessions into segment directories
+  await downloadScheduledSessions(firehoseBaseUrl, firehoseApiKey);
+
+  console.log('geoSegment() - finished');
+  process.exit(0);
```

Comment:
out of interest, did you include this because there still seemed to be some async processes running in the background (if so, that could be due to the `forEach(async` issue)?

Reply:
Yep that'll be it, deleted